Editor/EditorCore/DrawShapeTool.cs

```diff
@@ -0,0 +1,375 @@
+using System;
+using System.Linq;
+using UnityEditor.EditorTools;
+using UnityEngine;
+using UnityEngine.ProBuilder;
+using UnityEngine.ProBuilder.MeshOperations;
+using Math = UnityEngine.ProBuilder.Math;
+using UObject = UnityEngine.Object;
+
+namespace UnityEditor.ProBuilder
+{
+    class DrawShapeTool : EditorTool
+    {
+        enum InputState
+        {
+            SelectPlane,
+            SetWidthAndDepth,
+            SetHeight
+        }
+
+        [SerializeField]
+        InputState m_InputState;
+
+        ShapeComponent m_Shape;
+        SerializedObject m_Object;
+
+        // plane of interaction
+        Plane m_Plane;
+        Vector3 m_Forward;
+        Vector3 m_Right;
+        Vector3 m_Origin, m_OppositeCorner, m_HeightCorner;
+
+        Quaternion m_Rotation;
+        Bounds m_Bounds;
+
+        bool m_IsDragging;
+
+        GUIContent m_ShapeTitle;
+
+        static TypeCache.TypeCollection s_AvailableShapeTypes;
+        string[] m_ShapeTypesPopupContent;
+
+        static Pref<int> s_ActiveShapeIndex = new Pref<int>("ShapeBuilder.ActiveShapeIndex", 0);
+
+        static Vector3 s_Size;
+
+        static Type activeShapeType {
+            get { return s_ActiveShapeIndex < 0 ? typeof(Cube) : s_AvailableShapeTypes[s_ActiveShapeIndex]; }
+        }
+
+        static DrawShapeTool()
+        {
+            s_AvailableShapeTypes = TypeCache.GetTypesDerivedFrom<Shape>();
+        }
+
+        void OnEnable()
+        {
+            InitNewShape();
+            EditorTools.EditorTools.activeToolChanged += ActiveToolChanged;
+            m_ShapeTitle = new GUIContent("Draw Shape");
+            m_ShapeTypesPopupContent = s_AvailableShapeTypes.Select(x => x.Name).ToArray();
+        }
+
+        void InitNewShape()
+        {
+            m_Shape = new GameObject("Shape", typeof(ShapeComponent)).GetComponent<ShapeComponent>();
+            m_Shape.gameObject.hideFlags = HideFlags.HideAndDontSave;
+            m_Shape.hideFlags = HideFlags.None;
+            m_Shape.SetShape(CreateShape(activeShapeType));
+            m_Object = new SerializedObject(m_Shape);
+        }
+
+        void OnDisable()
+        {
+            if (m_Shape.gameObject.hideFlags == HideFlags.HideAndDontSave)
+            {
+                DestroyImmediate(m_Shape.gameObject);
+                InitNewShape();
+            }
+            EditorTools.EditorTools.activeToolChanged -= ActiveToolChanged;
```

Comment:
If I'm not mistaken, EditorTools.EditorTools will not be valid anymore in Unity 2020.2, you may use something as follows to handle the different unity versions :

```
#if UNITY_2020_2_OR_NEWER
using ToolManager = UnityEditor.EditorTools.ToolManager;
#else
using ToolManager = UnityEditor.EditorTools.EditorTools;
#endif

```